File: csharp/Dson/IO/BinaryUtils.cs
#region LICENSE

//  Copyright 2023 wjybxx
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System.Runtime.CompilerServices;
using Google.Protobuf;

namespace Dson.IO;

/// <summary>
/// C#10不支持逻辑右移，但这里使用算术右移是等价的
/// </summary>
public static class BinaryUtils
{
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static void CheckBuffer(byte[] buffer, int offset, int length) {
        CheckBuffer(buffer.Length, offset, length);
    }

    /// <summary>
    /// 检查buffer的参数
    /// </summary>
    /// <param name="bufferLength">buffer数组的长度</param>
    /// <param name="offset">数据的起始索引</param>
    /// <param name="length">数据的长度</param>
    /// <exception cref="ArgumentException"></exception>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static void CheckBuffer(int bufferLength, int offset, int length) {
        if ((offset | length | (bufferLength - (offset + length))) < 0) {
            throw new ArgumentException($"Array range is invalid. Buffer.length={bufferLength}, offset={offset}, length={length}");
        }
    }

    public static void CheckBuffer(int bufferLength, int offset) {
        if (offset < 0 || offset > bufferLength) {
            throw new ArgumentException($"Array range is invalid. Buffer.length={bufferLength}, offset={offset}");
        }
    }

    /** c#的byte默认是无符号的；这一点我觉得C#是对的... */
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static int ToUint(byte value) {
        return value;
    }

    #region 大端编码

    public static byte GetByte(byte[] buffer, int index) {
        return buffer[index];
    }

    public static void SetByte(byte[] buffer, int index, int value) {
        buffer[index] = (byte)value;
    }

    public static void SetShort(byte[] buffer, int index, int value) {
        buffer[index] = (byte)(value >> 8);
        buffer[index + 1] = (byte)value;
    }

    public static short GetShort(byte[] buffer, int index) {
        return (short)((buffer[index] << 8)
                       | (buffer[index + 1] & 0xff));
    }

    public static void SetInt(byte[] buffer, int index, int value) {
        buffer[index] = (byte)(value >> 24);
        buffer[index + 1] = (byte)(value >> 16);
        buffer[index + 2] = (byte)(value >> 8);
        buffer[index + 3] = (byte)value;
    }

    public static int GetInt(byte[] buffer, int index) {
        return (((buffer[index] & 0xff) << 24)
                | ((buffer[index + 1] & 0xff) << 16)
                | ((buffer[index + 2] & 0xff) << 8)
                | ((buffer[index + 3] & 0xff)));
    }

    public static void SetLong(byte[] buffer, int index, long value) {
        buffer[index] = (byte)(value >> 56);
        buffer[index + 1] = (byte)(value >> 48);
        buffer[index + 2] = (byte)(value >> 40);
        buffer[index + 3] = (byte)(value >> 32);
        buffer[index + 4] = (byte)(value >> 24);
        buffer[index + 5] = (byte)(value >> 16);
        buffer[index + 6] = (byte)(value >> 8);
        buffer[index + 7] = (byte)value;
    }

    public static long GetLong(byte[] buffer, int index) {
        return (((buffer[index] & 0xffL) << 56)
                | ((buffer[index + 1] & 0xffL) << 48)
                | ((buffer[index + 2] & 0xffL) << 40)
                | ((buffer[index + 3] & 0xffL) << 32)
                | ((buffer[index + 4] & 0xffL) << 24)
                | ((buffer[index + 5] & 0xffL) << 16)
                | ((buffer[index + 6] & 0xffL) << 8)
                | ((buffer[index + 7] & 0xffL)));
    }

    #endregion

    #region 小端编码

    public static void SetShortLe(byte[] buffer, int index, int value) {
        buffer[index] = (byte)value;
        buffer[index + 1] = (byte)(value >> 8);
    }

    public static short GetShortLe(byte[] buffer, int index) {
        return (short)((buffer[index] & 0xff)
                       | (buffer[index + 1] << 8));
    }

    public static void SetIntLe(byte[] buffer, int index, int value) {
        buffer[index] = (byte)value;
        buffer[index + 1] = (byte)(value >> 8);
        buffer[index + 2] = (byte)(value >> 16);
        buffer[index + 3] = (byte)(value >> 24);
    }

    public static int GetIntLe(byte[] buffer, int index) {
        return (((buffer[index] & 0xff))
                | ((buffer[index + 1] & 0xff) << 8)
                | ((buffer[index + 2] & 0xff) << 16)
                | ((buffer[index + 3] & 0xff) << 24));
    }

    public static void SetLongLe(byte[] buffer, int index, long value) {
        buffer[index] = (byte)value;
        buffer[index + 1] = (byte)(value >> 8);
        buffer[index + 2] = (byte)(value >> 16);
        buffer[index + 3] = (byte)(value >> 24);
        buffer[index + 4] = (byte)(value >> 32);
        buffer[index + 5] = (byte)(value >> 40);
        buffer[index + 6] = (byte)(value >> 48);
        buffer[index + 7] = (byte)(value >> 56);
    }

    public static long GetLongLe(byte[] buffer, int index) {
        return (((buffer[index] & 0xffL))
                | ((buffer[index + 1] & 0xffL) << 8)
                | ((buffer[index + 2] & 0xffL) << 16)
                | ((buffer[index + 3] & 0xffL) << 24)
                | ((buffer[index + 4] & 0xffL) << 32)
                | ((buffer[index + 5] & 0xffL) << 40)
                | ((buffer[index + 6] & 0xffL) << 48)
                | ((buffer[index + 7] & 0xffL) << 56));
    }

    #endregion

    #region protobuf decode

    // 由于Protobuf在C#端的接口开放较少，我决定不直接使用CodedInputStream和CodedOutputStream，而是自己实现

    /// <summary>
    /// 计算无符号32位整数(非负数)编码后的长度
    /// </summary>
    /// <param name="value"></param>
    /// <returns>编码长度</returns>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int ComputeUInt32Size(uint value) {
        return CodedOutputStream.ComputeRawVarint32Size(value);
    }

    /// <summary>
    /// 计算无符号64位整数(非负数)编码后的长度
    /// </summary>
    /// <param name="value"></param>
    /// <returns>编码长度</returns>
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int ComputeUint64Size(ulong value) {
        return CodedOutputStream.ComputeRawVarint64Size(value);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int ReadInt32(byte[] buffer, int pos, out int newPos) {
        ulong rawBits = ReadRawVarint64(buffer, pos, out newPos);
        return (int)rawBits;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static long ReadInt64(byte[] buffer, int pos, out int newPos) {
        ulong rawBits = ReadRawVarint64(buffer, pos, out newPos);
        return (int)rawBits;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int ReadUint32(byte[] buffer, int pos, out int newPos) {
        return (int)ReadRawVarint64(buffer, pos, out newPos);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static long ReadUint64(byte[] buffer, int pos, out int newPos) {
        return (long)ReadRawVarint64(buffer, pos, out newPos);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int ReadSint32(byte[] buffer, int pos, out int newPos) {
        ulong rawBits = ReadRawVarint64(buffer, pos, out newPos);
        return DecodeZigZag32((uint)rawBits);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static long ReadSint64(byte[] buffer, int pos, out int newPos) {
        ulong rawBits = ReadRawVarint64(buffer, pos, out newPos);
        return DecodeZigZag64(rawBits);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int ReadFixed32(byte[] buffer, int pos, out int newPos) {
        uint rawBits = ReadRawFixed32(buffer, pos, out newPos);
        return (int)rawBits;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static long ReadFixed64(byte[] buffer, int pos, out int newPos) {
        ulong rawBits = ReadRawFixed64(buffer, pos, out newPos);
        return (long)rawBits;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static float ReadFloat(byte[] buffer, int pos, out int newPos) {
        uint rawBits = ReadRawFixed32(buffer, pos, out newPos);
        return BitConverter.UInt32BitsToSingle(rawBits);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static double ReadDouble(byte[] buffer, int pos, out int newPos) {
        ulong rawBits = ReadRawFixed64(buffer, pos, out newPos);
        return BitConverter.UInt64BitsToDouble(rawBits);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static int DecodeZigZag32(uint n) => (int)(n >> 1) ^ -((int)n & 1);

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static long DecodeZigZag64(ulong n) => (long)(n >> 1) ^ -((long)n & 1L);

    /** varint编码不区分int和long，而是固定读取到高位字节为0，因此无需两个方法 */
    private static ulong ReadRawVarint64(byte[] buffer, int pos, out int newPos) {
        int shift = 0;
        ulong r = 0;
        do {
            byte b = buffer[pos++];
            r |= (b & 127UL) << shift; // 取后7位左移
            if (b < 128U) { // 高位0
                newPos = pos;
                return r;
            }
            shift += 7;
        } while (shift < 64);
        // 读取超过10个字节
        throw new DsonIOException("DsonInput encountered a malformed varint.");
    }

    private static uint ReadRawFixed32(byte[] buffer, int pos, out int newPos) {
        uint r = (((buffer[pos] & 0xffU))
                  | ((buffer[pos + 1] & 0xffU) << 8)
                  | ((buffer[pos + 2] & 0xffU) << 16)
                  | ((buffer[pos + 3] & 0xffU) << 24));
        newPos = pos + 4;
        return r;
    }

    private static ulong ReadRawFixed64(byte[] buffer, int pos, out int newPos) {
        ulong r = (((buffer[pos] & 0xffUL))
                   | ((buffer[pos + 1] & 0xffUL) << 8)
                   | ((buffer[pos + 2] & 0xffUL) << 16)
                   | ((buffer[pos + 3] & 0xffUL) << 24)
                   | ((buffer[pos + 4] & 0xffUL) << 32)
                   | ((buffer[pos + 5] & 0xffUL) << 40)
                   | ((buffer[pos + 6] & 0xffUL) << 48)
                   | ((buffer[pos + 7] & 0xffUL) << 56));
        newPos = pos + 8;
        return r;
    }

    #endregion

    #region protobuf encode

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int WriteInt32(byte[] buffer, int pos, int value) {
        if (value >= 0) {
            return WriteRawVarint32(buffer, pos, (uint)value);
        }
        else {
            return WriteRawVarint64(buffer, pos, (ulong)value);
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int WriteInt64(byte[] buffer, int pos, long value) {
        return WriteRawVarint64(buffer, pos, (ulong)value);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int WriteUint32(byte[] buffer, int pos, int value) {
        return WriteRawVarint32(buffer, pos, (uint)value);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int WriteUint64(byte[] buffer, int pos, long value) {
        return WriteRawVarint64(buffer, pos, (ulong)value);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int WriteSint32(byte[] buffer, int pos, int value) {
        return WriteRawVarint32(buffer, pos, EncodeZigZag32(value));
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int WriteSint64(byte[] buffer, int pos, long value) {
        return WriteRawVarint64(buffer, pos, EncodeZigZag64(value));
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int WriteFixed32(byte[] buffer, int pos, int value) {
        return WriteRawFixed32(buffer, pos, (uint)value);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int WriteFixed64(byte[] buffer, int pos, long value) {
        return WriteRawFixed64(buffer, pos, (ulong)value);
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int WriteFloat(byte[] buffer, int pos, float value) {
        return WriteRawFixed32(buffer, pos, BitConverter.SingleToUInt32Bits(value));
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    internal static int WriteDouble(byte[] buffer, int pos, double value) {
        return WriteRawFixed64(buffer, pos, BitConverter.DoubleToUInt64Bits(value));
    }

    /// <summary>
    /// 写入一个变长的64位整数，所有的负数都将固定10字节
    /// </summary>
    /// <param name="buffer"></param>
    /// <param name="pos">开始写入的位置</param>
    /// <param name="value">要写入的值</param>
    /// <returns>写入后的新坐标</returns>
    public static int WriteRawVarint64(byte[] buffer, int pos, ulong value) {
        if (value < 128UL) { // 小数值较多的情况下有意义
            buffer[pos] = (byte)value;
            return pos + 1;
        }
        while (true) {
            if (value > 127UL) {
                buffer[pos++] = (byte)((value & 127UL) | 128UL); // 截取后7位，高位补1
                value >>= 7;
            }
            else {
                buffer[pos++] = (byte)value;
                return pos;
            }
        }
    }

    /// <summary>
    /// 注意：该方法只可以在value为非负数的情况下可调用，外部在转换数据类型的时候要注意
    /// </summary>
    private static int WriteRawVarint32(byte[] buffer, int pos, uint value) {
        if (value < 128U) { // 小数值较多的情况下有意义
            buffer[pos] = (byte)value;
            return pos + 1;
        }
        while (true) {
            if (value > 127U) {
                buffer[pos++] = (byte)((value & 127U) | 128U); // 截取后7位，高位补1
                value >>= 7;
            }
            else {
                buffer[pos++] = (byte)value;
                return pos;
            }
        }
    }

    private static int WriteRawFixed32(byte[] buffer, int pos, uint value) {
        buffer[pos] = (byte)value;
        buffer[pos + 1] = (byte)(value >> 8);
        buffer[pos + 2] = (byte)(value >> 16);
        buffer[pos + 3] = (byte)(value >> 24);
        return pos + 4;
    }

    private static int WriteRawFixed64(byte[] buffer, int pos, ulong value) {
        buffer[pos] = (byte)value;
        buffer[pos + 1] = (byte)(value >> 8);
        buffer[pos + 2] = (byte)(value >> 16);
        buffer[pos + 3] = (byte)(value >> 24);
        buffer[pos + 4] = (byte)(value >> 32);
        buffer[pos + 5] = (byte)(value >> 40);
        buffer[pos + 6] = (byte)(value >> 48);
        buffer[pos + 7] = (byte)(value >> 56);
        return pos + 8;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static uint EncodeZigZag32(int n) => (uint)(n << 1 ^ n >> 31);

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private static ulong EncodeZigZag64(long n) => (ulong)(n << 1 ^ n >> 63);

    #endregion
}

File: java/Dson-Core/src/main/java/cn/wjybxx/dson/io/BinaryUtils.java
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson.io;

import org.apache.commons.lang3.mutable.MutableInt;

import java.nio.Buffer;

/**
 * @author wjybxx
 * date 2023/3/31
 */
public class BinaryUtils {

    private BinaryUtils() {
    }

    public static void checkBuffer(byte[] buffer, int offset, int length) {
        checkBuffer(buffer.length, offset, length);
    }

    /**
     * 允许{@code offset + length == bufferLength}
     *
     * @param bufferLength buffer数组的长度
     * @param offset       数据的起始索引
     * @param length       数据的长度
     */
    public static void checkBuffer(int bufferLength, int offset, int length) {
        if ((offset | length | (bufferLength - (offset + length))) < 0) {
            throw new IllegalArgumentException(String.format("Array range is invalid. Buffer.length=%d, offset=%d, length=%d",
                    bufferLength, offset, length));
        }
    }

    public static void checkBuffer(int bufferLength, int offset) {
        if (offset < 0 || offset > bufferLength) {
            throw new IllegalArgumentException(String.format("Array range is invalid. Buffer.length=%d, offset=%d",
                    bufferLength, offset));
        }
    }

    /** JDK9+的代码跑在JDK8上的兼容问题 */
    public static void position(Buffer byteBuffer, int newOffset) {
        byteBuffer.position(newOffset);
    }

    // region 大端编码

    public static byte getByte(byte[] buffer, int index) {
        return buffer[index];
    }

    public static void setByte(byte[] buffer, int index, int value) {
        buffer[index] = (byte) value;
    }

    public static void setShort(byte[] buffer, int index, int value) {
        buffer[index] = (byte) (value >>> 8);
        buffer[index + 1] = (byte) value;
    }

    public static short getShort(byte[] buffer, int index) {
        return (short) ((buffer[index] << 8)
                | (buffer[index + 1] & 0xff));
    }

    public static void setInt(byte[] buffer, int index, int value) {
        buffer[index] = (byte) (value >>> 24);
        buffer[index + 1] = (byte) (value >>> 16);
        buffer[index + 2] = (byte) (value >>> 8);
        buffer[index + 3] = (byte) value;
    }

    public static int getInt(byte[] buffer, int index) {
        return (((buffer[index] & 0xff) << 24)
                | ((buffer[index + 1] & 0xff) << 16)
                | ((buffer[index + 2] & 0xff) << 8)
                | ((buffer[index + 3] & 0xff)));
    }

    public static void setLong(byte[] buffer, int index, long value) {
        buffer[index] = (byte) (value >>> 56);
        buffer[index + 1] = (byte) (value >>> 48);
        buffer[index + 2] = (byte) (value >>> 40);
        buffer[index + 3] = (byte) (value >>> 32);
        buffer[index + 4] = (byte) (value >>> 24);
        buffer[index + 5] = (byte) (value >>> 16);
        buffer[index + 6] = (byte) (value >>> 8);
        buffer[index + 7] = (byte) value;
    }

    public static long getLong(byte[] buffer, int index) {
        return (((buffer[index] & 0xffL) << 56)
                | ((buffer[index + 1] & 0xffL) << 48)
                | ((buffer[index + 2] & 0xffL) << 40)
                | ((buffer[index + 3] & 0xffL) << 32)
                | ((buffer[index + 4] & 0xffL) << 24)
                | ((buffer[index + 5] & 0xffL) << 16)
                | ((buffer[index + 6] & 0xffL) << 8)
                | ((buffer[index + 7] & 0xffL)));
    }
    // endregion

    // region 小端编码

    public static void setShortLE(byte[] buffer, int index, int value) {
        buffer[index] = (byte) value;
        buffer[index + 1] = (byte) (value >>> 8);
    }

    public static short getShortLE(byte[] buffer, int index) {
        return (short) ((buffer[index] & 0xff)
                | (buffer[index + 1] << 8));
    }

    public static void setIntLE(byte[] buffer, int index, int value) {
        buffer[index] = (byte) value;
        buffer[index + 1] = (byte) (value >>> 8);
        buffer[index + 2] = (byte) (value >>> 16);
        buffer[index + 3] = (byte) (value >>> 24);
    }

    public static int getIntLE(byte[] buffer, int index) {
        return (((buffer[index] & 0xff))
                | ((buffer[index + 1] & 0xff) << 8)
                | ((buffer[index + 2] & 0xff) << 16)
                | ((buffer[index + 3] & 0xff) << 24));
    }

    public static void setLongLE(byte[] buffer, int index, long value) {
        buffer[index] = (byte) value;
        buffer[index + 1] = (byte) (value >>> 8);
        buffer[index + 2] = (byte) (value >>> 16);
        buffer[index + 3] = (byte) (value >>> 24);
        buffer[index + 4] = (byte) (value >>> 32);
        buffer[index + 5] = (byte) (value >>> 40);
        buffer[index + 6] = (byte) (value >>> 48);
        buffer[index + 7] = (byte) (value >>> 56);
    }

    public static long getLongLE(byte[] buffer, int index) {
        return (((buffer[index] & 0xffL))
                | ((buffer[index + 1] & 0xffL) << 8)
                | ((buffer[index + 2] & 0xffL) << 16)
                | ((buffer[index + 3] & 0xffL) << 24)
                | ((buffer[index + 4] & 0xffL) << 32)
                | ((buffer[index + 5] & 0xffL) << 40)
                | ((buffer[index + 6] & 0xffL) << 48)
                | ((buffer[index + 7] & 0xffL) << 56));
    }

    // endregion

    // 以下参考自protobuf，以避免引入PB
    // region protobuf util

    private static final int IntCodedMask1 = (-1) << 7; // 低7位0
    private static final int IntCodedMask2 = (-1) << 14; // 低14位0
    private static final int IntCodedMask3 = (-1) << 21;
    private static final int IntCodedMask4 = (-1) << 28;

    private static final long LongCodedMask1 = (-1L) << 7;
    private static final long LongCodedMask2 = (-1L) << 14;
    private static final long LongCodedMask3 = (-1L) << 21;
    private static final long LongCodedMask4 = (-1L) << 28;
    private static final long LongCodedMask5 = (-1L) << 35;
    private static final long LongCodedMask6 = (-1L) << 42;
    private static final long LongCodedMask7 = (-1L) << 49;
    private static final long LongCodedMask8 = (-1L) << 56;
    private static final long LongCodedMask9 = (-1L) << 63;

    /** 计算原始的32位变长整形的编码长度 */
    public static int computeRawVarInt32Size(int value) {
        if ((value & IntCodedMask1) == 0) return 1; // 所有高位为0
        if ((value & IntCodedMask2) == 0) return 2;
        if ((value & IntCodedMask3) == 0) return 3;
        if ((value & IntCodedMask4) == 0) return 4;
        return 5;
    }

    /** 计算原始的64位变长整形的编码长度 */
    public static int computeRawVarInt64Size(long value) {
        if ((value & LongCodedMask1) == 0) return 1; // 所有高位为0
        if ((value & LongCodedMask2) == 0) return 2;
        if ((value & LongCodedMask3) == 0) return 3;
        if ((value & LongCodedMask4) == 0) return 4;
        if ((value & LongCodedMask5) == 0) return 5;
        if ((value & LongCodedMask6) == 0) return 6;
        if ((value & LongCodedMask7) == 0) return 7;
        if ((value & LongCodedMask8) == 0) return 8;
        if ((value & LongCodedMask9) == 0) return 9;
        return 10;
    }

    public static int encodeZigZag32(int n) {
        return (n << 1) ^ (n >> 31);
    }

    public static int decodeZigZag32(final int n) {
        return (n >>> 1) ^ -(n & 1);
    }

    public static long encodeZigZag64(long n) {
        return (n << 1) ^ (n >> 63);
    }

    public static long decodeZigZag64(final long n) {
        return (n >>> 1) ^ -(n & 1);
    }
    // endregion

    //region protobuf decode

    public static int readInt32(byte[] buffer, int pos, MutableInt newPos) {
        long rawBits = readRawVarint64(buffer, pos, newPos);
        return (int) rawBits;
    }

    public static long readInt64(byte[] buffer, int pos, MutableInt newPos) {
        return readRawVarint64(buffer, pos, newPos);
    }

    public static int readUint32(byte[] buffer, int pos, MutableInt newPos) {
        return (int) readRawVarint64(buffer, pos, newPos);
    }

    public static long readUint64(byte[] buffer, int pos, MutableInt newPos) {
        return readRawVarint64(buffer, pos, newPos);
    }

    public static int readSint32(byte[] buffer, int pos, MutableInt newPos) {
        long rawBits = readRawVarint64(buffer, pos, newPos);
        return decodeZigZag32((int) rawBits);
    }

    public static long readSint64(byte[] buffer, int pos, MutableInt newPos) {
        long rawBits = readRawVarint64(buffer, pos, newPos);
        return decodeZigZag64(rawBits);
    }

    public static int readFixed32(byte[] buffer, int pos, MutableInt newPos) {
        return readRawFixed32(buffer, pos, newPos);
    }

    public static long readFixed64(byte[] buffer, int pos, MutableInt newPos) {
        return readRawFixed64(buffer, pos, newPos);
    }

    public static float readFloat(byte[] buffer, int pos, MutableInt newPos) {
        int rawBits = readRawFixed32(buffer, pos, newPos);
        return Float.intBitsToFloat(rawBits);
    }

    public static double readDouble(byte[] buffer, int pos, MutableInt newPos) {
        long rawBits = readRawFixed64(buffer, pos, newPos);
        return Double.longBitsToDouble(rawBits);
    }

    /** varint编码不区分int和long，而是固定读取到高位字节为0，因此无需两个方法 */
    private static long readRawVarint64(byte[] buffer, int pos, MutableInt newPos) {
        // 单字节优化
        byte b = buffer[pos++];
        long r = (b & 127L);
        if ((b & 128) == 0) {
            newPos.setValue(pos);
            return r;
        }
        int shift = 7;
        do {
            b = buffer[pos++];
            r |= (b & 127L) << shift; // 取后7位左移
            if ((b & 128) == 0) { // 高位0
                newPos.setValue(pos);
                return r;
            }
            shift += 7;
        } while (shift < 64);
        // 读取超过10个字节
        throw new DsonIOException("DsonInput encountered a malformed varint.");
    }

    private static int readRawFixed32(byte[] buffer, int pos, MutableInt newPos) {
        int r = (((buffer[pos] & 0xff))
                | ((buffer[pos + 1] & 0xff) << 8)
                | ((buffer[pos + 2] & 0xff) << 16)
                | ((buffer[pos + 3] & 0xff) << 24));
        newPos.setValue(pos + 4);
        return r;
    }

    private static long readRawFixed64(byte[] buffer, int pos, MutableInt newPos) {
        long r = (((buffer[pos] & 0xffL))
                | ((buffer[pos + 1] & 0xffL) << 8)
                | ((buffer[pos + 2] & 0xffL) << 16)
                | ((buffer[pos + 3] & 0xffL) << 24)
                | ((buffer[pos + 4] & 0xffL) << 32)
                | ((buffer[pos + 5] & 0xffL) << 40)
                | ((buffer[pos + 6] & 0xffL) << 48)
                | ((buffer[pos + 7] & 0xffL) << 56));
        newPos.setValue(pos + 8);
        return r;
    }

    //endregion

    //region protobuf encode

    /** @return newPos */
    public static int writeInt32(byte[] buffer, int pos, int value) {
        if (value >= 0) {
            return writeRawVarint32(buffer, pos, value);
        } else {
            return writeRawVarint64(buffer, pos, value);
        }
    }

    public static int writeInt64(byte[] buffer, int pos, long value) {
        return writeRawVarint64(buffer, pos, value);
    }

    public static int writeUint32(byte[] buffer, int pos, int value) {
        return writeRawVarint32(buffer, pos, value);
    }

    public static int writeUint64(byte[] buffer, int pos, long value) {
        return writeRawVarint64(buffer, pos, value);
    }

    public static int writeSint32(byte[] buffer, int pos, int value) {
        return writeRawVarint32(buffer, pos, encodeZigZag32(value));
    }

    public static int writeSint64(byte[] buffer, int pos, long value) {
        return writeRawVarint64(buffer, pos, encodeZigZag64(value));
    }

    public static int writeFixed32(byte[] buffer, int pos, int value) {
        return writeRawFixed32(buffer, pos, value);
    }

    public static int writeFixed64(byte[] buffer, int pos, long value) {
        return writeRawFixed64(buffer, pos, value);
    }

    public static int writeFloat(byte[] buffer, int pos, float value) {
        return writeRawFixed32(buffer, pos, Float.floatToRawIntBits(value));
    }

    public static int writeDouble(byte[] buffer, int pos, double value) {
        return writeRawFixed64(buffer, pos, Double.doubleToRawLongBits(value));
    }

    /**
     * 写入一个变长的64位整数，所有的负数都将固定10字节
     *
     * @param buffer buffer
     * @param pos    开始写入的位置
     * @param value  要写入的值
     * @return 写入后的新坐标
     */
    private static int writeRawVarint64(byte[] buffer, int pos, long value) {
        if (value >= 0 && value < 128L) { // 小数值较多的情况下有意义
            buffer[pos] = (byte) value;
            return pos + 1;
        }
        while (true) {
            if ((value & LongCodedMask1) != 0) {
                buffer[pos++] = (byte) ((value & 127L) | 128L); // 截取后7位，高位补1
                value >>>= 7; // java必须逻辑右移
            } else {
                buffer[pos++] = (byte) value;
                return pos;
            }
        }
    }

    /**
     * 写入一个变长的32位整数，负数将固定为5字节
     * （注意：普通int慎重调用，这里将int看做无符号整数编码）
     *
     * @param buffer buffer
     * @param pos    开始写入的位置
     * @param value  要写入的值
     * @return 写入后的新坐标
     */
    private static int writeRawVarint32(byte[] buffer, int pos, int value) {
        if (value >= 0 && value < 128) { // 小数值较多的情况下有意义
            buffer[pos] = (byte) value;
            return pos + 1;
        }
        while (true) {
            if ((value & IntCodedMask1) != 0) {
                buffer[pos++] = (byte) ((value & 127) | 128); // 截取后7位，高位补1
                value >>>= 7; // java必须逻辑右移
            } else {
                buffer[pos++] = (byte) value;
                return pos;
            }
        }
    }

    private static int writeRawFixed32(byte[] buffer, int pos, int value) {
        buffer[pos] = (byte) value;
        buffer[pos + 1] = (byte) (value >> 8);
        buffer[pos + 2] = (byte) (value >> 16);
        buffer[pos + 3] = (byte) (value >> 24);
        return pos + 4;
    }

    private static int writeRawFixed64(byte[] buffer, int pos, long value) {
        buffer[pos] = (byte) value;
        buffer[pos + 1] = (byte) (value >> 8);
        buffer[pos + 2] = (byte) (value >> 16);
        buffer[pos + 3] = (byte) (value >> 24);
        buffer[pos + 4] = (byte) (value >> 32);
        buffer[pos + 5] = (byte) (value >> 40);
        buffer[pos + 6] = (byte) (value >> 48);
        buffer[pos + 7] = (byte) (value >> 56);
        return pos + 8;
    }

    //endregion
}